x-pack/plugins/enterprise_search/public/applications/app_search/components/relevance_tuning/relevance_tuning.test.tsx

```diff
@@ -5,33 +5,150 @@
+    expect(buttons.length).toBe(2);
+    const saveButton = shallow(buttons[0]);
+    saveButton.simulate('click');
+    expect(actions.updateSearchSettings).toHaveBeenCalled();
+  });
+
+  it('renders a Reset button that will remove all weights and boosts', () => {
+    const buttons = subject().find(EuiPageHeader).prop('rightSideItems') as React.ReactElement[];
+    expect(buttons.length).toBe(2);
+    const resetButton = shallow(buttons[1]);
+    resetButton.simulate('click');
+    expect(actions.resetSearchSettings).toHaveBeenCalled();
+  });
+
+  describe('when the engine has no schema', () => {
+    let wrapper: ShallowWrapper;
+
+    beforeAll(() => {
+      // An eninge would have no schema if it is newly created, and no documents have been indexed
+      // yet.
+      setMockValues({
+        ...values,
+        engineHasSchemaFields: false,
+      });
+      wrapper = subject();
+    });
+
+    it('will not render buttons if the engine has no schema', () => {
+      setMockValues({
+        ...values,
+        engineHasSchemaFields: false,
+      });
+      const buttons = wrapper.find(EuiPageHeader).prop('rightSideItems') as React.ReactElement[];
+      expect(buttons.length).toBe(0);
+    });
+
+    it('will render an empty message', () => {
+      setMockValues({
+        ...values,
+        engineHasSchemaFields: false,
+      });
+      expect(wrapper.find(EuiEmptyPrompt).exists()).toBe(true);
+      expect(wrapper.find(RelevanceTuningForm).exists()).toBe(false);
+    });
+  });
+
+  it('shows a message when there are invalid boosts', () => {
+    // An invalid boost would be if a user creats a functional boost on a number field, then that
+    // field later changes to text. At this point, the boost still exists but is invalid for
+    // a text field.
+    setMockValues({
+      ...values,
+      engine: {
+        invalidBoosts: true,
+        unsearchedUnconfirmedFields: false,
+      },
+    });
+    expect(subject().find('[data-test-subj="RelevanceTuningInvalidBoostsCallout"]').exists()).toBe(
+      true
+    );
+  });
+
+  it('shows a message when there are unconfirmed fields', () => {
+    // An invalid boost would be if a user creats a functional boost on a number field, then that
+    // field later changes to text. At this point, the boost still exists but is invalid for
+    // a text field.
+    setMockValues({
+      ...values,
+      engine: {
+        invalidBoosts: false,
+        unsearchedUnconfirmedFields: true,
+      },
+    });
+    expect(
+      subject().find('[data-test-subj="RelevanceTuningUnsearchedFieldsCallout"]').exists()
+    ).toBe(true);
+  });
+
+  it('shows a message when there are schema field conflicts', () => {
+    // Schema conflicts occur when a meta engine has fields in source engines with have differing types,
+    // hence relevance tuning cannot be applied as we don't know the actual type
```

Comment:
I'm a super huge fan of these comments in the test files - they really feel like they're embracing the concept of unit tests being living documentation!

Reply:
Right???? I like as well.